### DISCUSSION TITLE
DIY FCB1010 clone build questions

### DISCUSSION THREAD

**Discussion by Solone** [tttapa/Control-Surface#917] [General]
Im new to using this library and I'm attempting building a DIY clone of the Behringer FCB1010 foot controller to control a software looper called Sooperlooper, but only sending note messages from the buttons. 

My question is this: Sometimes I see people using the begin() method on NoteButtons to check the state as the NoteButton object uses the built in pull-up resistor, but then I see other code just seems to rely on the USBMidi_Interface update() method.

Can I just declare my objects  and their settings and rely on the update() method to check it all? Is it that simple?

Thanks in advance

**Comment by tttapa** [comment]
> Sometimes I see people using the begin() method on NoteButtons to check the state as the NoteButton object uses the built in pull-up resistor,

Could you link to an example of this?

---

When you call `Control_Surface.begin()`, it will automatically call the `begin()` function for most* objects that have been created up to that point and have not been disabled. A similar story for `Control_Surface.loop()`, it will call the `update()` function for objects that are enabled.

(*) You can find the implementation of `Control_Surface.begin()` here:

https://github.com/tttapa/Control-Surface/blob/39182bbab79487236c11579db3acfa336813bf54/src/Control_Surface/Control_Surface_Class.cpp#L22-L45

As you can see, it initializes all Extended IO elements (like multiplexers and shift registers), all MIDI interfaces, all MIDI outputs, all displays, all MIDI Input Elements, and all generic Updatables (which includes all MIDI Output Elements, including `NoteButton`).

There are cases in which you don't _need_ all of that, for example, when you just use a single MIDI interface and no MIDI Output Elements, you can just initialize and update the MIDI interface directly, without going through `Control_Surface.begin/loop()`. This is done in some of the examples in the MIDI tutorial, for example: https://tttapa.github.io/Control-Surface-doc/Doxygen/d3/df7/midi-tutorial.html

---

In conclusion, if you create all your NoteButtons, MIDI interfaces, and other objects up front, you can rely on the `Control_Surface.begin/loop` functions to initialize and update them correctly.  
Only in rare cases where you need to dynamically add/remove or enable/disable objects could it be that you have to call the `begin/update` functions for those objects manually.

**Comment by Solone** [comment]
Hi! Thanks for the comprehensive answer. My confusion it seems has been with exactly what you describe as being done globally (so to speak) from Control_Surface.begin() and Control_Surface.update() as against specific directly from beginning and enabling objects. 

I'm still fairly new to all this so just working with note buttons right now. 
What am I doing wrong with the transposer in my code? 
I seem to just get glitching note on/note off messages??
My friend and I checked all switches and soldering and it seems ok??

```cpp
#include <Control_Surface.h>

USBMIDI_Interface midi; 

void setup() {
  Control_Surface.begin();
}

// Instantiate a Transposer that can transpose from one octave down to one
// octave up
Transposer<-12, +12> transposer;
 
// Instantiate a Selector to change the transposition
IncrementDecrementSelector<transposer.getNumberOfBanks()> selector {
  transposer,
  {12, 13},
  Wrap::Clamp,
};
 
//velocity
const uint8_t velocity = 0x7F;

// Instantiate an array of NoteButton objects
Bankable::NoteButton buttons[] {
  {transposer, 0, MIDI_Notes::C(4)}, 
  {transposer, 1, MIDI_Notes::Db(4)},
  {transposer, 2, MIDI_Notes::D(4)},
  {transposer, 3, MIDI_Notes::Eb(4)}, 
  {transposer, 4, MIDI_Notes::E(4)},
  {transposer, 5, MIDI_Notes::F_(4)},
  {transposer, 6, MIDI_Notes::Gb(4)}, 
  {transposer, 7, MIDI_Notes::G(4)},
  {transposer, 8, MIDI_Notes::Ab(4)},
  {transposer, 9, MIDI_Notes::A(4)}, 
  {transposer, 10, MIDI_Notes::B(4)},
  {transposer, 11, MIDI_Notes::C(5)}
};

/**
CCButton But1 = { 8, 62};
CCButton But2 = { 4, 63};
CCButton But3 = { 5, 64};
CCButton But5 = { 6, 65};
CCButton But6 = { 2, 66};
CCButton But8 = { 3, 67};
CCButton But9 = { 10, 68};
CCButton But10 = { 11, 69};
**/

void loop() {
  Control_Surface.update();
}
```

Thanks a ton for the help so far.

**Reply by Solone** [reply]
Ok, seconds after I posted my reply, I tried compiling and realised I had to change Control.Surface.update() to Control_Surface.loop(). Thats one thing fixed but it still gives me an endless stream of noteOn (C#4 127) noteOff (C#4 127) when I press button 1 on pin 0 ???

**Reply by tttapa** [reply]
Which board are you using? Some boards use pins 0/1 for the UART.  
How are your buttons wired? They should be connected between the GPIO pins and ground.

The code looks alright, apart from the fact that this line has no effect:
> `const uint8_t velocity = 0x7F;`

**Reply by Solone** [reply]
Hi Piéter,
I'm using an arduino Uno R3 with the 16U2 midi controller chip reprogrammed with the mocolufa dual firmware. My buttons are all wired beginning with pin 0 to 13 and the common to GND

**Reply by tttapa** [reply]
You'll have to use different pins: pins 0 and 1 are the UART pins connecting the ATmega328P to the ATmega16U2. When using MIDI, you cannot use these pins for anything else.

**Comment by Solone** [comment]
OK Thanks, I'll try a 4067 multiplexer to get all my needed buttons!

**Reply by tttapa** [reply]
Keep in mind that you can use pins 14-19 (A0-A5) as digital pins as well.

**Reply by Solone** [reply]
Awesome! Ok that's new to me! Forgive my ignorance but how would I do that?

**Reply by tttapa** [reply]
There's no difference: the “analog” pins A0-A5 are just standard digital pins that happen to be connected to the analog-to-digital converter as an extra feature. You can just connect your buttons to them like you're doing now, and then simply change the pin number in you sketch to e.g. `A0` instead of `0`. (`A0` is just a constant with the value `14`.)
